tests/lib/prepare.sh

```diff
@@ -456,90 +456,75 @@ prepare_classic() {
 }
 
-repack_kernel_snap() {
-    local TARGET=$1
-    local VERSION
-    local UNPACK_DIR
-    local CHANNEL
-
-    VERSION=$(nested_get_version)
-    if [ "$VERSION" = 16 ]; then
-        CHANNEL=latest
-    else
-        CHANNEL=$VERSION
+run_snapcraft() {
+    ensure_snapcraft
+    if ! (cd "${PROJECT_PATH}" && snapcraft "$@"); then
+        # shellcheck disable=SC2012
+        tail -n1000 "${HOME}/.local/state/snapcraft/log/$(ls -t "${HOME}/.local/state/snapcraft/log/" -1 | head -n1)"
+        false
     fi
+    cleanup_snapcraft
+}
 
-    echo "Repacking kernel snap"
-    UNPACK_DIR=/tmp/kernel-unpack
-    snap download --basename=pc-kernel --channel="$CHANNEL/${KERNEL_CHANNEL}" pc-kernel
-    unsquashfs -no-progress -d "$UNPACK_DIR" pc-kernel.snap
-    snap pack --filename="$TARGET" "$UNPACK_DIR"
+build_snapd_snap() {
+    local TARGET
+    local snapd_snap_cache
+    TARGET="${1}"
+
+    snapd_snap_cache="${NESTED_WORK_DIR:-/tmp/work-dir}/snapd_snap"
+    [ -d "${snapd_snap_cache}" ] || mkdir -p "${snapd_snap_cache}"
+    for snap in "${snapd_snap_cache}"/snapd_*.snap; do
+        if ! [ -f "${snap}" ]; then
+            if os.query is-pc-amd64 && cp "${PROJECT_PATH}/built-snap"/snapd_1337.*.snap.keep "${snapd_snap_cache}/snapd_from_ci.snap"; then
+                break
+            fi
+            [ -d "${TARGET}" ] || mkdir "${TARGET}"
+            chmod -R go+r "${PROJECT_PATH}/tests"
+            run_snapcraft --use-lxd --verbosity quiet --output="snapd_from_snapcraft.snap"
+            mv "${PROJECT_PATH}"/snapd_from_snapcraft.snap "${snapd_snap_cache}"
+        fi
+        break
+    done
 
-    rm -rf pc-kernel.snap "$UNPACK_DIR"
+    cp "${snapd_snap_cache}"/snapd_*.snap "${TARGET}/"
 }
 
-repack_snapd_snap_with_deb_content_and_run_mode_firstboot_tweaks() {
-    local TARGET="$1"
+build_snapd_snap_with_run_mode_firstboot_tweaks() {
+    local snapd_snap_cache
+    local TARGET
 
-    local UNPACK_DIR="/tmp/snapd-unpack"
-    unsquashfs -no-progress -d "$UNPACK_DIR" snapd_*.snap
+    TARGET="${1}"
 
-    # data/preseed.json is not included in the deb, use the latest
-    # version from source tree to replace the one in the re-packed snapd snap.
-    cp "$PROJECT_PATH/data/preseed.json" "$UNPACK_DIR"/usr/lib/snapd
+    snapd_snap_cache="${NESTED_WORK_DIR:-/tmp/work-dir}/snapd_snap_with_tweaks"
+    [ -d "${snapd_snap_cache}" ] || mkdir -p "${snapd_snap_cache}"
+    for snap in "${snapd_snap_cache}"/snapd_*.snap; do
+        if [ -f "${snap}" ]; then
+            cp "${snap}" "${TARGET}/"
+            return
+        fi
+    done
 
-    # clean snap apparmor.d to ensure we put the right snap-confine apparmor
-    # file in place. Its called usr.lib.snapd.snap-confine on 14.04 but
-    # usr.lib.snapd.snap-confine.real everywhere else
-    rm -f "$UNPACK_DIR"/etc/apparmor.d/*
+    if ! os.query is-pc-amd64 || ! cp "${PROJECT_PATH}/built-snap"/snapd_1337.*.snap.keep "/tmp/snapd_from_snapcraft.snap"; then
```

Comment:
it is not clear why: 
! cp "${PROJECT_PATH}/built-snap"/snapd_1337.*.snap.keep "/tmp/snapd_from_snapcraft.snap"

Reply:
If running outside the github workflow, `"${PROJECT_PATH}/built-snap"` will not exist. `cp` will fail. And then we fallback to building the snap.

Reply:
I'd rather have an explicit environment variable controlling whether the build through snapcraft is (dis)allowed, otherwise there's a chance that everything will seem to be set up correctly, but the prepare will keep building snapd with snapcraft. , eg `TESTS_USE_PREBUILT_SNAPD_SNAP: "$(HOST: echo PREBUILT_SNAPD)"` in spread.yaml and then fail hard here?

Reply:
That makes total sense.